File: eztalk_proxy/services/web_search.py
import asyncio
import logging
import orjson
from typing import List, Dict

from googleapiclient.discovery import build
from googleapiclient.errors import HttpError

from ..core.config import GOOGLE_API_KEY_ENV, GOOGLE_CSE_ID, SEARCH_RESULT_COUNT, SEARCH_SNIPPET_MAX_LENGTH

logger = logging.getLogger("EzTalkProxy.WebSearch")

async def perform_web_search(query: str, rid: str) -> List[Dict[str, str]]:
    logger.info(f"RID-{rid}: perform_web_search called. Query: '{query}'. GOOGLE_API_KEY_ENV is set: {bool(GOOGLE_API_KEY_ENV)}, GOOGLE_CSE_ID is set: {bool(GOOGLE_CSE_ID)}")
    results = []
    actual_google_api_key = GOOGLE_API_KEY_ENV
    if not actual_google_api_key or not GOOGLE_CSE_ID:
        logger.warning(f"RID-{rid}: Web search skipped. GOOGLE_API_KEY or GOOGLE_CSE_ID not set in environment variables.")
        logger.warning(f"RID-{rid}: To enable web search, please create a .env file in the root directory of the 'backdAiTalk' project and add your Google API Key and Custom Search Engine ID.")
        logger.warning(f"RID-{rid}: Example .env file content:\n# Google Custom Search API Key\nGOOGLE_API_KEY=\"YOUR_GOOGLE_API_KEY\"\n# Google Custom Search Engine ID\nGOOGLE_CSE_ID=\"YOUR_GOOGLE_CSE_ID\"")
        return results
    if not query:
        logger.warning(f"RID-{rid}: Web search skipped, query is empty.")
        return results

    try:
        def search_sync():
            service = build("customsearch", "v1", developerKey=actual_google_api_key, cache_discovery=False)
            res = service.cse().list(q=query, cx=GOOGLE_CSE_ID, num=min(SEARCH_RESULT_COUNT, 10)).execute()
            return res.get('items', [])

        logger.info(f"RID-{rid}: Performing web search for query: '{query[:100]}'")
        search_items = await asyncio.to_thread(search_sync)

        for i, item in enumerate(search_items):
            snippet = item.get('snippet', 'N/A').replace('\n', ' ').strip()
            if len(snippet) > SEARCH_SNIPPET_MAX_LENGTH:
                snippet = snippet[:SEARCH_SNIPPET_MAX_LENGTH] + "..."
            results.append({
                "index": i + 1,
                "title": item.get('title', 'N/A').strip(),
                "href": item.get('link', 'N/A'),
                "snippet": snippet
            })
        logger.info(f"RID-{rid}: Web search completed, found {len(results)} results.")

    except HttpError as e:
        err_content = "Unknown Google API error"
        status_code = "N/A"
        if hasattr(e, 'resp') and hasattr(e.resp, 'status'):
            status_code = e.resp.status
        try:
            content_json = orjson.loads(e.content)
            err_detail = content_json.get("error", {})
            err_message = err_detail.get("message", str(e.content))
            err_content = f"{err_message} (Code: {err_detail.get('code', 'N/A')}, Status: {err_detail.get('status', 'N/A')})"
        except:
            err_content = e._get_reason() if hasattr(e, '_get_reason') else e.content.decode(errors='ignore')[:200]
        logger.error(f"RID-{rid}: Google Web Search HttpError (Status: {status_code}): {err_content}")
    except Exception as search_exc:
        logger.error(f"RID-{rid}: Google Web Search failed for query '{query[:50]}': {search_exc}", exc_info=True)
    return results

def generate_search_context_message_content(query: str, search_results: List[Dict[str, str]]) -> str:
    logger.debug(f"generate_search_context_message_content: Processing query '{query}' with {len(search_results)} results")
    
    if not search_results:
        logger.debug("generate_search_context_message_content: No search results, returning empty string")
        return ""
    
    # 使用更紧凑的格式，减少不必要的换行符
    prompt_parts = [
        f"You have been provided with the following web search results for the user's query: '{query}'. "
        "Your task is to synthesize this information, along with your general knowledge, to construct a comprehensive and natural-sounding answer. "
        "It is crucial that you DO NOT include any inline citation marks like [1], [2], [Source 1], etc., directly in your response text. "
        "The user will have a separate way to view the sources if they wish."
    ]

    # 将所有搜索结果合并为一个更紧凑的格式
    sources_text = "Search Results: "
    total_snippet_length = 0
    
    for i, res in enumerate(search_results):
        source_identifier = res.get('index', i + 1)
        title = res.get('title', 'N/A').strip()
        snippet = res.get('snippet', 'N/A').strip()
        
        # 去除snippet中的多余换行符和空白
        original_snippet = snippet
        snippet = ' '.join(snippet.split())
        total_snippet_length += len(snippet)
        
        if snippet != original_snippet:
            logger.debug(f"generate_search_context_message_content: Cleaned snippet {i+1}: {len(original_snippet)} -> {len(snippet)} chars")
        
        sources_text += f"[{source_identifier}] {title}: {snippet}. "
    
    logger.info(f"generate_search_context_message_content: Total snippet content length: {total_snippet_length} chars")
    
    prompt_parts.append(sources_text.strip())
    
    prompt_parts.append(
        "Based on the information from these sources and your existing knowledge, please formulate your answer. "
        "Focus on delivering a clear, accurate, and well-integrated response to the user's query. "
        "Remember, do not insert any citation markers (e.g., [1], [Source 2]) into the body of your answer."
    )
    
    # 使用单个换行符连接，避免产生大量空行
    final_content = " ".join(prompt_parts)
    
    logger.info(f"generate_search_context_message_content: Generated search context of {len(final_content)} chars")
    logger.debug(f"generate_search_context_message_content: Context preview: {final_content[:200]}...")
    
    # 检查是否包含可能导致空白段落的模式
    newline_count = final_content.count('\n')
    if newline_count > 5:
        logger.warning(f"generate_search_context_message_content: High newline count detected: {newline_count}")
    
    double_space_count = final_content.count('  ')
    if double_space_count > 0:
        logger.warning(f"generate_search_context_message_content: Double spaces detected: {double_space_count}")
    
    return final_content

File: src/services/web-search.js
/**
 * Web search service for Cloudflare Workers
 * Matches backend-docker web search functionality
 */

import { logger } from '../utils/logger';

export class WebSearchService {
  constructor(env) {
    this.env = env;
    this.googleApiKey = env.GOOGLE_API_KEY;
    this.googleCseId = env.GOOGLE_CSE_ID;
    this.searchResultCount = parseInt(env.SEARCH_RESULT_COUNT || '5');
    this.snippetMaxLength = parseInt(env.SEARCH_SNIPPET_MAX_LENGTH || '200');
  }

  /**
   * Perform web search using Google Custom Search API
   */
  async performWebSearch(query, requestId) {
    const logPrefix = `RID-${requestId}`;
    logger.info(`${logPrefix}: perform_web_search called. Query: '${query}'. API Key set: ${!!this.googleApiKey}, CSE ID set: ${!!this.googleCseId}`);

    const results = [];

    if (!this.googleApiKey || !this.googleCseId) {
      logger.warn(`${logPrefix}: Web search skipped. GOOGLE_API_KEY or GOOGLE_CSE_ID not set in environment variables.`);
      logger.warn(`${logPrefix}: To enable web search, please set GOOGLE_API_KEY and GOOGLE_CSE_ID in Cloudflare Workers environment variables.`);
      return results;
    }

    if (!query || query.trim() === '') {
      logger.warn(`${logPrefix}: Web search skipped, query is empty.`);
      return results;
    }

    try {
      const searchUrl = new URL('https://www.googleapis.com/customsearch/v1');
      searchUrl.searchParams.set('key', this.googleApiKey);
      searchUrl.searchParams.set('cx', this.googleCseId);
      searchUrl.searchParams.set('q', query);
      searchUrl.searchParams.set('num', Math.min(this.searchResultCount, 10).toString());

      logger.info(`${logPrefix}: Performing web search for query: '${query.substring(0, 100)}'`);

      const response = await fetch(searchUrl.toString());
      
      if (!response.ok) {
        const errorText = await response.text();
        logger.error(`${logPrefix}: Google Custom Search API error: ${response.status} - ${errorText}`);
        return results;
      }

      const data = await response.json();
      const items = data.items || [];

      for (let i = 0; i < items.length; i++) {
        const item = items[i];
        let snippet = (item.snippet || 'N/A').replace(/\n/g, ' ').trim();
        
        if (snippet.length > this.snippetMaxLength) {
          snippet = snippet.substring(0, this.snippetMaxLength) + '...';
        }

        results.push({
          index: i + 1,
          title: (item.title || 'N/A').trim(),
          href: item.link || 'N/A',
          snippet: snippet
        });
      }

      logger.info(`${logPrefix}: Web search completed, found ${results.length} results.`);

    } catch (error) {
      logger.error(`${logPrefix}: Web search error:`, error);
    }

    return results;
  }

  /**
   * Generate search context message content
   */
  generateSearchContextMessageContent(searchResults, query, requestId) {
    const logPrefix = `RID-${requestId}`;
    
    if (!searchResults || searchResults.length === 0) {
      logger.info(`${logPrefix}: No search results to include in context.`);
      return '';
    }

    logger.info(`${logPrefix}: Generating search context with ${searchResults.length} results.`);

    let contextContent = `\n\n--- Web Search Results for "${query}" ---\n`;
    
    for (const result of searchResults) {
      contextContent += `\n${result.index}. **${result.title}**\n`;
      contextContent += `   URL: ${result.href}\n`;
      contextContent += `   Summary: ${result.snippet}\n`;
    }
    
    contextContent += '\n--- End of Web Search Results ---\n\n';
    contextContent += 'Please use the above search results to provide accurate and up-to-date information in your response. ';
    contextContent += 'Cite the sources when appropriate using the provided URLs.\n\n';

    return contextContent;
  }

  /**
   * Check if web search should be performed
   */
  shouldPerformWebSearch(requestData) {
    // Check if web search is explicitly enabled
    if (requestData.use_web_search === true || requestData.qwen_enable_search === true) {
      return true;
    }

    // Check if the query contains search-related keywords
    const lastMessage = requestData.messages[requestData.messages.length - 1];
    if (!lastMessage || lastMessage.role !== 'user') {
      return false;
    }

    const content = lastMessage.content || 
                   (lastMessage.parts && lastMessage.parts.find(p => p.type === 'text_content')?.text) || '';
    
    const searchKeywords = [
      'search for', 'find information', 'latest news', 'current events',
      'what happened', 'recent developments', 'up to date', 'latest update'
    ];

    return searchKeywords.some(keyword => 
      content.toLowerCase().includes(keyword.toLowerCase())
    );
  }
}